File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int height;
    static int width;
    
    public int solution(int[][] maps) {
        
        height = maps.length;
        width = maps[0].length;
        bfs(maps);
        
        int answer = maps[height-1][width-1];
        return answer == 1 ? -1 : answer;
    }
    
    public static void bfs(int[][] maps){
        int[] dirY = {+1, 0, -1, 0};
        int[] dirX = {0, +1, 0, -1};
        
        Queue<int[]> que = new LinkedList<>();
        que.add(new int[] {0, 0});
        
        while(!que.isEmpty()){
            int[] now = que.poll();
            
            for(int i=0; i<4; i++){
                int newY = now[0] + dirY[i];
                int newX = now[1] + dirX[i];
                
                if(newY < 0 || newY >= height || newX < 0 || newX >= width) continue;
                if(maps[newY][newX] == 0 || maps[newY][newX] > 1) continue;
                
                maps[newY][newX] = maps[now[0]][now[1]] + 1;
                que.add(new int[] {newY, newX});
            }
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
const dirY = [-1, 0, +1, 0];
const dirX = [0, +1, 0, -1];
let board;
let visited;
let N, M;

function bfs() {
    const que = [];
    que.push([0, 0]);
    visited[0][0] = true;
    
    while(que.length){
        const now = que.shift();
        
        for(let i=0; i<4; i++){
            const newY = now[0] + dirY[i];
            const newX = now[1] + dirX[i];
            
            if(newY < 0 || newX < 0 
               || newY >= N || newX >= M) continue;
            if(board[newY][newX] === 0 || visited[newY][newX]) continue;
            
            que.push([newY, newX]);
            visited[newY][newX] = true;
            board[newY][newX] = board[now[0]][now[1]]+1;
        }
    }
}

function solution(maps) {
    board = maps;
    N = board.length;
    M = board[0].length;
    visited = Array.from({length: N}, () => new Array(M).fill(false));
    
    bfs();
    
    return board[N-1][M-1] === 1 ? -1 : board[N-1][M-1];
}